File: sepa-base/src/main/java/com/adyen/checkout/sepa/SepaOutputData.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 22/8/2019.
 */

package com.adyen.checkout.sepa;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.TextUtils;

import com.adyen.checkout.base.component.OutputData;
import com.adyen.checkout.base.validation.ValidatedField;

class SepaOutputData implements OutputData {

    private final ValidatedField<String> mOwnerNameField;
    private final ValidatedField<String> mIbanNumberField;
    private final Iban mIban;

    SepaOutputData(@NonNull String ownerName, @NonNull String ibanNumber) {
        mOwnerNameField = new ValidatedField<>(ownerName,
                TextUtils.isEmpty(ownerName) ? ValidatedField.Validation.PARTIAL : ValidatedField.Validation.VALID);
        mIban = Iban.parse(ibanNumber);
        mIbanNumberField = validateIbanNumber(ibanNumber, mIban);
    }

    @Override
    public boolean isValid() {
        return mOwnerNameField.isValid() && mIbanNumberField.isValid();
    }

    @NonNull
    public ValidatedField<String> getOwnerNameField() {
        return mOwnerNameField;
    }

    @NonNull
    public ValidatedField<String> getIbanNumberField() {
        return mIbanNumberField;
    }

    @Nullable
    public Iban getIban() {
        return mIban;
    }

    private ValidatedField<String> validateIbanNumber(@NonNull String ibanNumber, @Nullable Iban iban) {
        final ValidatedField.Validation validation;
        if (iban != null) {
            validation = ValidatedField.Validation.VALID;
        } else if (Iban.isPartial(ibanNumber)) {
            validation = ValidatedField.Validation.PARTIAL;
        } else {
            validation = ValidatedField.Validation.INVALID;
        }

        return new ValidatedField<>(ibanNumber, validation);
    }
}

File: sepa/src/main/java/com/adyen/checkout/sepa/SepaOutputData.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 22/8/2019.
 */
package com.adyen.checkout.sepa

import com.adyen.checkout.components.base.OutputData
import com.adyen.checkout.components.ui.FieldState
import com.adyen.checkout.components.ui.Validation
import com.adyen.checkout.sepa.Iban.Companion.parse

class SepaOutputData(ownerName: String, ibanNumber: String) : OutputData {

    val ownerNameField: FieldState<String> = FieldState(
        value = ownerName,
        validation = if (ownerName.isEmpty())
            Validation.Invalid(R.string.checkout_holder_name_not_valid)
        else
            Validation.Valid
    )
    val ibanNumberField: FieldState<String>
    val iban: Iban? = parse(ibanNumber)

    override val isValid: Boolean
        get() = ownerNameField.validation.isValid() && ibanNumberField.validation.isValid()

    private fun validateIbanNumber(ibanNumber: String, iban: Iban?): FieldState<String> {
        val validation: Validation = if (iban != null) {
            Validation.Valid
        } else {
            Validation.Invalid(R.string.checkout_iban_not_valid)
        }
        return FieldState(ibanNumber, validation)
    }

    init {
        ibanNumberField = validateIbanNumber(ibanNumber, iban)
    }
}
